### ISSUE TITLE
Stale RunnerReplicasets are not deleted after updating the spec of RunnerDeployment

### ISSUE THREAD

**Issue by aylei** [actions-runner-controller/actions-runner-controller#1642]
### Controller Version

0.25.1

### Helm Chart Version

0.20.1

### CertManager Version

v1.8.2

### Deployment Method

Helm

### cert-manager installation

Yes

### Checks

- [X] This isn't a question or user support case (For Q&A and community support, go to [Discussions](https://github.com/actions-runner-controller/actions-runner-controller/discussions). It might also be a good idea to contract with any of contributors and maintainers if your business is so critical and therefore you need priority support
- [X] I've read [releasenotes](https://github.com/actions-runner-controller/actions-runner-controller/tree/master/docs/releasenotes) before submitting this issue and I'm sure it's not due to any recently-introduced backward-incompatible changes
- [X] My actions-runner-controller version (v0.x.y) does support the feature
- [X] I've already upgraded ARC (including the CRDs, see charts/actions-runner-controller/docs/UPGRADING.md for details) to the latest and it didn't fix the issue

### Resource Definitions

```yaml
apiVersion: actions.summerwind.dev/v1alpha1
kind: RunnerDeployment
metadata:
  name: arm64-runner
  namespace: cicd
spec:
  effectiveTime: "2022-07-13T05:23:36Z"
  replicas: 0
  selector: null
  template:
    metadata: {}
    spec:
      dockerdContainerResources: {}
      image: ""
      labels:
      - arm64-runner
      - eks
      nodeSelector:
        beta.kubernetes.io/arch: arm64
      organization: matrixorigin
      resources:
        limits:
          cpu: "2"
          memory: 7Gi
        requests:
          cpu: "1"
          memory: 2Gi
      serviceAccountName: cicd
---
apiVersion: actions.summerwind.dev/v1alpha1
kind: HorizontalRunnerAutoscaler
metadata:
  name: github-runner-arm64
  namespace: cicd
spec:
  capacityReservations:
  - effectiveTime: "2022-07-13T05:23:24Z"
    expirationTime: "2022-07-13T05:28:24Z"
    replicas: 1
  - effectiveTime: "2022-07-13T05:23:36Z"
    expirationTime: "2022-07-13T05:28:36Z"
    replicas: 1
  maxReplicas: 12
  minReplicas: 0
  scaleTargetRef:
    name: arm64-runner
  scaleUpTriggers:
  - amount: 1
    duration: 5m0s
    githubEvent:
      checkRun:
        status: queued
        types:
        - created
```

### To Reproduce

```markdown
1. Create an `RunnerDeployment`
2. Trigger scale-out via github event
3. Update the `RunnerDeployment`
4. The old replicaset would not be removed nor managed
```

### Describe the bug

Old `RunnerReplicaset` is not deleted after updating the spec of `RunnerDeployment`.

### Describe the expected behavior

After updating the spec of `RunnerDeployment`, the old `RunnerReplicaSet` should be removed once the workflows they were running are done.

### Controller Logs

```shell
> kg rdeploy
NAME           DESIRED   CURRENT   UP-TO-DATE   AVAILABLE   AGE
arm64-runner   0         0         0            0           19h
> kg runnerreplicaset
NAME                 DESIRED   CURRENT   READY   AGE
arm64-runner-22qdt   0         0         0       149m
arm64-runner-brc5c   1         1         1       18h
arm64-runner-dxsbr   2         2         2       4h46m
> k logs -f arc-actions-runner-controller-7777659cb4-hp8zr  | grep arm64-runner-dxsbr
(no logs about the stale RunnerReplicaSet)
```

### Runner Pod Logs

```shell
N/A
```

**Comment by mumoshu** [comment]
@aylei Hey! Thanks for reporting.
Can we rephrase this as "Resetting rdeploy's replicas from non-zero to zero doesn't remove runnerreplicasets", as that seems like what you've done?
Or did you actually see this behavior even if you didn't reset the replicas to zero? (What if you omit `replicas` from the RunnerDeployment spec on first creation and on update, so that spec changes are made for fields other than `replicas`?

**Comment by aylei** [comment]
I omitted the `replicas` field of `RunnerDeployment` and actually it was the `HorizontalRunnerAutoscaler` that set the `replias` to 0 (since all webhook events were expired), but the stale `RunnerReplicaset` (which was created by old `RunnerDeployment` spec were not scaled-in nor deleted)

> @aylei Hey! Thanks for reporting. Can we rephrase this as "Resetting rdeploy's replicas from non-zero to zero doesn't remove runnerreplicasets", as that seems like what you've done? Or did you actually see this behavior even if you didn't reset the replicas to zero? (What if you omit `replicas` from the RunnerDeployment spec on first creation and on update, so that spec changes are made for fields other than `replicas`?

**Comment by mumoshu** [comment]
@aylei Thanks for clarifying! I guess this might be a bug since 0.25.0, although I'm still unsure which change resulted in it.
Have you used ARC before? Was a previous version of ARC running fine in your cluster?

**Comment by mumoshu** [comment]
@aylei Hey! Could you share the full log from ARC's controller-manager? It would contain important information that is necessary to further drill it down to the cause.

**Comment by mumoshu** [comment]
A bit more context on this- the expected behavior in your scenario is that it firstly sets the desired replicas on older runerreplicasets to `0`. ARC starts removing runners and their runner pods for those older runnerreplicasets. Once the runners and runner pods get terminated, runnerreplicasets `CURRENT` replicas go down to `0`. Finally, the older runnerreplicasets are deleted hence disappear from the kubectl get output.

Seeing you report, the desired replicas of the older runnerreplicasets, `arm64-runner-brc5c` and `arm64-runner-dxsbr`, are not even updated to `0`. I'm still wondering how that could happen, and probably ARC's controller-manager logs help investigating it.

**Comment by peimanja** [comment]
Having the same issue after upgrading the controller in one of our clusters. You can see desired replicas on older runnerreplicasets are also not set to 0. We manually deleted the old ones and will check if this happens again

```
kubectl get runnerreplicaset
NAME                                DESIRED   CURRENT   READY   AGE
default-generic-lg-runner-b5sm4     0         0         0       19h
default-generic-lg-runner-z76wq     0         0         0       9d
default-generic-md-runner-4d58t     0         0         0       19h
default-generic-md-runner-dc5tt     0         0         0       9d
default-generic-nano-runner-vpcnn   0         0         0       9d
default-generic-nano-runner-zvsfs   0         0         0       19h
default-generic-sm-runner-klxqm     0         0         0       9d
default-generic-sm-runner-nmxp7     0         0         0       19h
default-generic-xl-runner-47wss     0         0         0       19h
default-generic-xl-runner-mpcqz     0         0         0       9d
default-generic-xs-runner-p4zws     5         5         5       9d
default-generic-xs-runner-vllj4     0         0         0       19h
infra-generic-lg-runner-6qgbr       0         0         0       19h
infra-generic-lg-runner-pfcjl       1         1         1       9d
infra-generic-md-runner-9s52l       0         0         0       19h
infra-generic-md-runner-ks8qh       0         0         0       9d
infra-generic-nano-runner-9m2kk     0         0         0       19h
infra-generic-nano-runner-js5qx     0         0         0       9d
infra-generic-sm-runner-fchbs       0         0         0       19h
infra-generic-sm-runner-rt7k6       0         0         0       9d
infra-generic-xl-runner-t57wv       1         1         1       19h
infra-generic-xl-runner-v64nl       0         0         0       9d
infra-generic-xs-runner-hth4x       0         0         0       19h
infra-generic-xs-runner-sx9wf       0         0         0       9d
```

**Comment by mumoshu** [comment]
@peimanja Hey! Thanks for reporting.
Could you fill in the bug report, perhaps in a new issue, so that I might be one step closer to reproducing this? (I'm still unable to reproduce this myself. My runnerreplicasets just go away as soon as it becomes unnecessary.

**Comment by mumoshu** [comment]
@peimanja I'm especially interested in how your RunnerDeployment specs look like. But a full bug report would be even more awesome.

**Comment by mumoshu** [comment]
@peimanja @aylei We've fixed a bug related to runnerreplicaset in 0.25.2. I'm still unsure if it's relevant to this one, but it would be great if you could upgrade your ARC instances to v0.25.2 and see if it fixes the issue.

**Comment by peimanja** [comment]
@mumoshu So today I've upgraded our other cluster using `0.25.2` and everything including `runnerreplicaset` were handled correctly. Looks like you somehow fixed the issue :)

**Comment by aylei** [comment]
`v0.25.1`
```
kg runnerreplicaset -n cicd
NAME                  DESIRED   CURRENT   READY   AGE
pulumi-runner-5px7z   0         0         0       2m19s
pulumi-runner-vdmjg   0         0         0       11d
```

`v0.25.2`
```
kg runnerreplicaset -n cicd
NAME                  DESIRED   CURRENT   READY   AGE
pulumi-runner-vn4c6   0         0         0       72s
```

@mumoshu I also verified `v0.25.2` in my environment and the orphaned `runnerreplicaset` are properly cleaned in the new version. So I think this issue has been addressed, thanks!

**Comment by mumoshu** [comment]
@peimanja @aylei Thanks for your confirmation! Linking #1651 as the fix for this issue, which is included since v0.25.2.
